solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -632,19 +636,24 @@ private void refreshCollectionList(Watcher watcher) throws KeeperException, Inte
   }
 
   private class LazyCollectionRef extends ClusterState.CollectionRef {
-
     private final String collName;
+    private long lastUpdateTime;
+    private DocCollection cachedDocCollection;
 
     public LazyCollectionRef(String collName) {
       super(null);
       this.collName = collName;
+      this.lastUpdateTime = -1;
     }
 
     @Override
-    public DocCollection get() {
+    public synchronized DocCollection get() {
```

Comment:
I thought synchronized here would provide a good enough performance increase without the complexity of other approches